src/compute/src/server.rs

```diff
@@ -399,63 +401,39 @@ pub async fn compute_node_serve(
-                {
-                    AwaitTreeMiddlewareLayer::new_optional(await_tree_reg).layer(srv)
-                }
-            })
-            .add_service(MonitorServiceServer::new(monitor_srv))
-            .add_service(ConfigServiceServer::new(config_srv))
-            .add_service(HealthServer::new(health_srv))
-            .monitored_serve_with_shutdown(
-                listen_addr,
-                "grpc-compute-node-service",
-                TcpConfig {
-                    tcp_nodelay: true,
-                    keepalive_duration: None,
-                },
-                async move {
-                    tokio::select! {
-                        _ = tokio::signal::ctrl_c() => {},
-                        _ = &mut shutdown_recv => {
-                            for (join_handle, shutdown_sender) in sub_tasks {
-                                if let Err(_err) = shutdown_sender.send(()) {
-                                    tracing::warn!("Failed to send shutdown");
-                                    continue;
-                                }
-                                if let Err(err) = join_handle.await {
-                                    tracing::warn!(error = %err.as_report(), "Failed to join shutdown");
-                                }
-                            }
-                        },
-                    }
-                },
-            )
-            .await;
-    });
-    join_handle_vec.push(join_handle);
+    let server = tonic::transport::Server::builder()
+        .initial_connection_window_size(MAX_CONNECTION_WINDOW_SIZE)
+        .initial_stream_window_size(STREAM_WINDOW_SIZE)
+        .http2_max_pending_accept_reset_streams(Some(config.server.grpc_max_reset_stream as usize))
+        .layer(TracingExtractLayer::new())
+        // XXX: unlimit the max message size to allow arbitrary large SQL input.
+        .add_service(TaskServiceServer::new(batch_srv).max_decoding_message_size(usize::MAX))
+        .add_service(ExchangeServiceServer::new(exchange_srv).max_decoding_message_size(usize::MAX))
+        .add_service({
+            let await_tree_reg = stream_srv.mgr.await_tree_reg().cloned();
+            let srv = StreamServiceServer::new(stream_srv).max_decoding_message_size(usize::MAX);
+            #[cfg(madsim)]
+            {
+                srv
+            }
+            #[cfg(not(madsim))]
+            {
+                AwaitTreeMiddlewareLayer::new_optional(await_tree_reg).layer(srv)
+            }
+        })
+        .add_service(MonitorServiceServer::new(monitor_srv))
+        .add_service(ConfigServiceServer::new(config_srv))
+        .add_service(HealthServer::new(health_srv))
+        .monitored_serve_with_shutdown(
+            listen_addr,
+            "grpc-compute-node-service",
+            TcpConfig {
+                tcp_nodelay: true,
+                keepalive_duration: None,
+            },
+            shutdown.clone().cancelled_owned(),
```

Comment:
Directly pass shutdown token here instead of listening to Ctrl-C signal separately in each worker node.